tct/src/internal/hash/cache.rs

```diff
@@ -0,0 +1,104 @@
+//! A thread-safe cache intended hold lazily evaluated hashes.
+
+use std::{cell::Cell, fmt::Debug};
+
+use parking_lot::Once;
+
+use crate::Hash;
+
+/// A write-once cache for `Hash`es, allowing lazy evaluation of hashes inside the tree.
+#[derive(Default)]
+pub struct CachedHash {
+    /// A cell containing a cached hash value, which may not yet be set.
+    ///
+    /// IMPORTANT: It is **unsafe** to **read** or **write** this cell without synchronizing on the
+    /// `Once` within this struct.
+    cell: Cell<Hash>,
+    /// A synchronization variable which we will use to ensure that it is set at most once, ever, in
+    /// the lifetime of this `CachedHash` (thus allowing this struct to be marked `Sync`), and that
+    /// the value of the `Cell` is only ever observed after it has been set.
+    once: Once,
+}
+
+// This is safe because the restricted API of `CachedHash` only sets the internal hash *at most
+// once*, from only one thread, using a `Once` to ensure multiple threads do not race on setting the
+// hash. The only way to do interior mutability is via `set_if_empty`, which mutates the inner
+// `Cell` exactly once. Additionally, the value of the inner `Cell` cannot be observed until after
+// it has been fully set, which means it is impossible to observe partial writes, regardless of
+// whether this would be possible in practice on any supported architecture.
+unsafe impl Sync for CachedHash {}
```

Comment:
We should absolutely not be doing an `unsafe impl Sync` in our Merkle tree implementation.  It is critical that the Merkle tree work correctly, and having to reason about synchronization behaviour on top of basic functional correctness adds a huge amount of risk for very little gain.

Reply:
This is now resolved in the tct-integration-tct-only branch.